File: backend/models/recipe.py
from enum import StrEnum

from pydantic import BaseModel, Field


class UnitEnum(StrEnum):
    """Enum for units of measurement."""

    G = "g"
    KG = "kg"
    ML = "ml"
    L = "l"
    PCS = "pcs"
    TBSP = "tbsp"
    TSP = "tsp"


class Ingredient(BaseModel):
    """An ingredient model."""

    name: str
    unit: UnitEnum
    amount: float
    group: str | None = None


class Step(BaseModel):
    """A step model."""

    instructions: str


class RecipeBase(BaseModel):
    """The base recipe model (used when creating a recipe)."""

    title: str
    description: str | None = None
    portions: int
    ingredients: list[Ingredient]
    cooking_time: int
    steps: list["RecipeStep"]
    categories: list[str]
    gallery_images: list[int] | None = None
    cover_image: int | None = None


class LLMRecipe(RecipeBase):
    description: str | None = Field(
        description="A short description of the recipe. Focus on the meal not original recipe text, keep it one sentence and describe it in the simplest way."
    )
    gallery_image_urls: list[str] = Field(
        description="A list of recipe image urls contained in the data that should be included in the recipe."
    )
    categories: list[str] = Field(
        description="A list of recipe categories that should be included in the recipe. Focus on the most important labels. Less is more!"
    )
    is_a_recipe: bool = Field(
        description="True if the data is a recipe for food or a drink, False otherwise."
    )


class Recipe(RecipeBase):
    """A recipe with all attributes."""

    id_: int = -1
    creator_name: str | None = None
    creator_id: int | None = None
    clicks: int | None = None


class RecipeListing(BaseModel):
    """A slimmer recipe model for listings."""

    id_: int = -1
    title: str
    creator: str | None = None
    description: str
    categories: list[str]
    cover_image: int | None = None
    rating: float | None = None
    clicks: int | None = None
    cooking_time: int


class RecipeStep(BaseModel):
    """A recipe step model."""

    order_id: int
    step: str
    images: list[int] | None = None


class ImageID(BaseModel):
    """An image model with an id."""

    id_: int = -1

File: gobackend/internal/data/recipe.go
package data

type Unit string

const (
	UnitG    Unit = "g"
	UnitKG   Unit = "kg"
	UnitML   Unit = "ml"
	UnitL    Unit = "l"
	UnitPCS  Unit = "pcs"
	UnitTBSP Unit = "tbsp"
	UnitTSP  Unit = "tsp"
)

type Category string

const (
	CategoryMain       Category = "Hauptgericht"
	CategorySide       Category = "Beilage"
	CategoryBread      Category = "Brot"
	CategorySauce      Category = "Sauce"
	CategoryVegetarian Category = "Vegetarisch"
	CategoryVegan      Category = "Vegan"
	CategoryDrink      Category = "Getränk"
	CategoryBaked      Category = "Gebäck"
	CategoryAsian      Category = "Asia"
)

type Ingredient struct {
	Name   string  `json:"name"`
	Unit   Unit    `json:"unit"`
	Amount float64 `json:"amount"`
	Group  *string `json:"group,omitempty"`
}

type RecipeStep struct {
	OrderID int    `json:"order_id"`
	Step    string `json:"step"`
	Images  []int  `json:"images,omitempty"`
}

type Recipe struct {
	ID            int          `json:"id"`
	Title         string       `json:"title"`
	Description   *string      `json:"description,omitempty"`
	Portions      int          `json:"portions"`
	Ingredients   []Ingredient `json:"ingredients"`
	CookingTime   int          `json:"cooking_time"`
	Steps         []RecipeStep `json:"steps"`
	Categories    []Category   `json:"categories"`
	GalleryImages []int        `json:"gallery_images,omitempty"`
	CoverImage    *int         `json:"cover_image,omitempty"`
	CreatorName   *string      `json:"creator_name,omitempty"`
	CreatorID     *int         `json:"creator_id,omitempty"`
	Clicks        *int         `json:"clicks,omitempty"`
}

type RecipeListing struct {
	ID          int        `json:"id"`
	Title       string     `json:"title"`
	Creator     *string    `json:"creator,omitempty"`
	Description string     `json:"description"`
	Categories  []Category `json:"categories"`
	CoverImage  *int       `json:"cover_image,omitempty"`
	Rating      *float64   `json:"rating,omitempty"`
	Clicks      *int       `json:"clicks,omitempty"`
	CookingTime int        `json:"cooking_time"`
}
